REGRESSION (281882@main): LayoutTests/fast/repaint/content-visibility-hidden-container-with-abs-pos-c-v-auto-child.html fails.

mattwoodrow opened the pull request:
#### 6ee5bd84aa85a20079fbf8e63c42439c7a9b5cec
<pre>
REGRESSION (281882@main): LayoutTests/fast/repaint/content-visibility-hidden-container-with-abs-pos-c-v-auto-child.html fails.
<a href="https://bugs.webkit.org/show_bug.cgi?id=290723">https://bugs.webkit.org/show_bug.cgi?id=290723</a>
&lt;<a href="https://rdar.apple.com/148183112">rdar://148183112</a>&gt;

Reviewed by Alan Baradlay.

Changing the value of content-visibility requires that child layers recompute
their visible content status (RenderLayer::computeHasVisibleContent() checks
isSkippedContent(), not hasSkippedContent()).

We also can&apos;t use the fast path of setHasVisibleContent() (instead of setting
the dirty bit) when setting visibility:visible, since the layer can still be
hidden (by content-visibility, or overflow truncation).

Duplicate the existing repaint rect test to have a reftest equivalent.

* LayoutTests/fast/repaint/content-visibility-hidden-container-with-abs-pos-c-v-auto-child-reftest-expected.html: Added.
* LayoutTests/fast/repaint/content-visibility-hidden-container-with-abs-pos-c-v-auto-child-reftest.html: Added.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange):

Canonical link: <a href="https://commits.webkit.org/293085@main">https://commits.webkit.org/293085@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c708e75e5e9cc2db8a3e742d08cf8bd24cda7de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97768 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17393 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7610 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102880 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48297 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99813 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17687 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25846 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74462 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31646 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100771 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13417 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88378 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54814 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13193 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6296 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47740 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83186 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6373 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/105149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24848 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18109 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/105149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25220 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84540 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/105149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20929 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27525 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5204 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18455 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24809 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29978 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24631 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27945 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26205 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -900,18 +900,15 @@ void RenderElement::styleWillChange(StyleDifference diff, const RenderStyle& new
-        if (wasVisible != willBeVisible) {
+        if ((m_style.usedVisibility() == Visibility::Visible) != (newStyle.usedVisibility() == Visibility::Visible)) {
+            if (CheckedPtr layer = enclosingLayer())
+                layer->dirtyVisibleContentStatus();
```
we don't worry about over invalidation here? The content may still have the same perceived visibility when going from `visibility: visible` to `visibility: hidden` when inside skipped content?

mattwoodrow replied:
This is only dirtying the 'is visible' flag, so that we recompute visibility status next time we update descendant dependant flags. That may very well end up not being a change, and not cause any further invalidation of expensive things.

That said, unrelated to this code, changing `visibility` causes a `StyleDifference::Repaint` hint, so the problem still exists. I guess all the various types of visibility should be combined into a `StyleDifferenceContextSensitiveProperty` so that we can compute the final state and only invalidate as necessary.

alanbaradlay commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -900,18 +900,15 @@ void RenderElement::styleWillChange(StyleDifference diff, const RenderStyle& new
-                } else if (layer->hasVisibleContent() && (this == &layer->renderer() || layer->renderer().style().usedVisibility() != Visibility::Visible))
-                    layer->dirtyVisibleContentStatus();
+                for (RenderLayer* child = layer->firstChild(); child; child = child->nextSibling())
+                    child->dirtyVisibleContentStatus();
```
any reason not marking the enclosing layer dirty? e.g. the entire layer goes hidden when content-visibility changes from 'visible' to 'hidden' (`m_style.hasSkippedContent() != newStyle.hasSkippedContent()` is true)

mattwoodrow replied:
Changed this a bit, once I understood the `isSkippedRootOrSkippedContent()` a bit better.

It now just marks the layer directly, if the used value of content-visibility directly (and since that's inherited for hidden subtrees, marks all the actually-hidden children layers).

In theory we don't need to mark the skipped-root as having dirty visible content, but it looks like we can't easily distinguish those from this caller.

mattwoodrow committed:
REGRESSION (281882@main): LayoutTests/fast/repaint/content-visibility-hidden-container-with-abs-pos-c-v-auto-child.html fails.
https://bugs.webkit.org/show_bug.cgi?id=290723
<rdar://148183112>

Reviewed by Alan Baradlay.

Changing the value of content-visibility requires that child layers recompute
their visible content status (RenderLayer::computeHasVisibleContent() checks
isSkippedContent(), not hasSkippedContent()).

We also can't use the fast path of setHasVisibleContent() (instead of setting
the dirty bit) when setting visibility:visible, since the layer can still be
hidden (by content-visibility, or overflow truncation).

Duplicate the existing repaint rect test to have a reftest equivalent.

* LayoutTests/fast/repaint/content-visibility-hidden-container-with-abs-pos-c-v-auto-child-reftest-expected.html: Added.
* LayoutTests/fast/repaint/content-visibility-hidden-container-with-abs-pos-c-v-auto-child-reftest.html: Added.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange):

Canonical link: https://commits.webkit.org/293085@main
